ipa-core/src/protocol/ipa_prf/boolean_ops/multiplication.rs

```diff
@@ -1,33 +1,158 @@
-use std::iter::zip;
-
+use crate::protocol::basics::mul::SecureMul;
 use crate::{
     error::Error,
     ff::boolean::Boolean,
     protocol::{
-        context::Context,
-        RecordId,
-        ipa_prf::boolean_ops::{
-            addition_sequential::integer_add,
-        },
+        boolean::NBitStep, context::Context,
+        ipa_prf::boolean_ops::addition_sequential::integer_add, BooleanProtocols, RecordId,
     },
     secret_sharing::{replicated::semi_honest::AdditiveShare, BitDecomposed, FieldSimd},
 };
 
-
+/// This function multiplies x by y in these steps:
+/// 1. Double the input precision for y and repeat the most significant bit in the extra bits (Sign extension)
+///    X is assumed to be a positive number, so we will automatically pad with ZERO.
+/// 2. Repeatedly multiply x with each digits of y, shift the result 1 digit up each time
+/// 3. Add up the partial products using integer_add
+/// x is assumed to be a positive number
+/// y is assumed to be in two's complement and can be either signed or unsigned
+#[allow(dead_code)]
 pub async fn integer_mul<C, S, const N: usize>(
     ctx: C,
     record_id: RecordId,
     x: &BitDecomposed<AdditiveShare<Boolean, N>>,
     y: &BitDecomposed<AdditiveShare<Boolean, N>>,
-) -> Result<
-    BitDecomposed<AdditiveShare<Boolean, N>>,
-    Error,
->
+) -> Result<BitDecomposed<AdditiveShare<Boolean, N>>, Error>
 where
     C: Context,
     S: NBitStep,
     Boolean: FieldSimd<N>,
     AdditiveShare<Boolean, N>: BooleanProtocols<C, N>,
 {
-    //TODO: To be implemented
+    let new_len = x.len() + y.len();
+    let mut y = y.clone();
+    y.resize(new_len, y[y.len() - 1].clone());
+
+    let mut result = BitDecomposed::with_capacity(new_len);
+    for (i, yb) in y.into_iter().enumerate() {
+        let ctx_for_bit_of_y = ctx.narrow(&S::from(i));
+        let product_of_x_and_yb = ctx_for_bit_of_y
+            .parallel_join(x.iter().take(new_len - i).enumerate().map(|(j, xb)| {
+                let ctx_for_x_times_y_combo = ctx_for_bit_of_y.narrow(&S::from(j));
+                let yb = yb.clone();
+                async move { yb.multiply(xb, ctx_for_x_times_y_combo, record_id).await }
+            }))
+            .await?;
+
+        let mut t = BitDecomposed::with_capacity(new_len);
+        // |x| * y(i) << i
+        t.resize(i, AdditiveShare::ZERO);
```

Comment:
This will incur additional cost at the aggregation phase, since it will require more carries than necessary.

Instead of this, we can just leave the least significant `i` bits of `result` alone (we know they will be unchanged), and just invoke `integer_add` on the `new_len - i` most significant bits of `result` and `t`.